plugins/discovery-file/src/main/java/org/elasticsearch/discovery/file/FileBasedUnicastHostsProvider.java

```diff
@@ -0,0 +1,110 @@
+import java.nio.file.NoSuchFileException;
+import java.nio.file.Path;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+
+/**
+ * An implementation of {@link UnicastHostsProvider} that reads hosts/ports
+ * from {@link #UNICAST_HOSTS_FILE}.
+ *
+ * Each unicast host/port that is part of the discovery process must be listed on
+ * a separate line.  If the port is left off an entry, a default port of 9300 is
+ * assumed.  An example unicast hosts file could read:
+ *
+ * 67.81.244.10
+ * 67.81.244.11:9305
+ * 67.81.244.15:9400
+ */
+public class FileBasedUnicastHostsProvider extends AbstractComponent implements UnicastHostsProvider {
+
+    static final String UNICAST_HOSTS_FILE = "unicast_hosts.txt";
+
+    private final TransportService transportService;
+
+    private final Path unicastHostsFilePath;
+
+    @Inject
+    public FileBasedUnicastHostsProvider(Settings settings, TransportService transportService) {
+        super(settings);
+        this.transportService = transportService;
+        this.unicastHostsFilePath = new Environment(settings).configFile().resolve(UNICAST_HOSTS_FILE);
+    }
+
+    @Override
+    public List<DiscoveryNode> buildDynamicNodes() {
+        List<String> hostsList;
+        try {
+            hostsList = Files.readAllLines(unicastHostsFilePath);
+        } catch (FileNotFoundException | NoSuchFileException e) {
+            logger.warn((Supplier<?>) () -> new ParameterizedMessage("[discovery-file] Failed to find unicast hosts file [{}]",
+                                                                        unicastHostsFilePath), e);
+            return Collections.emptyList();
+        } catch (IOException e) {
+            logger.warn((Supplier<?>) () -> new ParameterizedMessage("[discovery-file] Error reading unicast hosts file [{}]",
+                                                                        unicastHostsFilePath), e);
+            return Collections.emptyList();
+        }
+
+        final List<DiscoveryNode> discoNodes = new ArrayList<>();
+        final int hostsSize = hostsList.size();
+        for (int i = 0; i < hostsSize; i++) {
+            final String host = hostsList.get(i);
+            TransportAddress[] addresses;
+            try {
+                addresses = transportService.addressesFromString(host, 1);
+            } catch (Exception e) {
+                logger.warn((Supplier<?>) () -> new ParameterizedMessage("[discovery-file] Failed to parse transport address from [{}]",
+                                                                            host), e);
+                return Collections.emptyList();
+            }
+            if (addresses.length != 1) {
+                logger.warn((Supplier<?>) () -> new ParameterizedMessage("[discovery-file] Failed to parse transport address from [{}]",
+                                                                            host));
+                return Collections.emptyList();
```

Comment:
> That said, I'd be in favor of throwing exceptions here and in the other places instead

See my comment above. We can throw exceptions here, but we'll just catch it higher up (and log it). We also ping based on the cluster state, so we can let the pinging continue and see how it goes.